text/java/0000-apache-tomcat-launch-config.md

```diff
@@ -0,0 +1,60 @@
+# Apache Tomcat launch configuration
+
+## Summary
+
+With the current implementation of the Apache Tomcat buildpack, its quite difficult to configure values at launch time that affect values in the `web.xml`, `server.xml` & `context.xml`. We 
+want to provide a way to make some features easily configurable and controlled at launch time rather than build time.
+
+## Motivation
+
+The Apache Tomcat buildpack provides a method of adding a configuration overlay using the build time `BP_TOMCAT_EXT_CONF_*` variables, this takes a pre-built archive
+and extracts this during build time. This method can be cumbersome if there are values that need to be changed on a per-deployment basis.  Apache Tomcat already supports
+the use of [system property replacements](https://tomcat.apache.org/tomcat-9.0-doc/config/systemprops.html#Property_replacements) within its configuration files.  The aim is to make
+the configuring of the system properties at launch time lot easier.
+ 
+## Detailed Explanation
+
+Given an example `context.xml`, which can be included in either an external archive, or inside the applications `META-INF` directory:
+
+```
+<Context>
+    <Resources allowLinking="true"/>
+
+    <Resource name="jdbc/Datasource" auth="Container"
+              type="javax.sql.DataSource" driverClassName="org.postgresql.Driver"
+              url="${postgres.url}" username="${postgres.username}"
+              password="${postgres.password}" maxTotal="25"
+              maxIdle="10" validationQuery="select 1"/>
+
+</Context>
+```
+
+The resource configuration is static apart from the `url`, `username` and `password` parameters, but we do not want these to be written into the image at build time.
+
+## Rationale and Alternatives
+
+It is possible to use `BPE_*` parameters with the [environment variables buildpack](https://github.com/paketo-buildpacks/environment-variables), these are written into the image at buildtime.
+
+These values could also be configured by updating `JAVA_TOOL_OPTIONS` variable, but this can be quite cumbersome and error prone if there are multiple values to configure, or if this env var already
+contains configuration.
+
+## Implementation
+
+The implementation will add a helper to the tomcat buildpack that will look for launch variables named `BPL_TOMCAT_ENV_*`, e.g. `BPL_TOMCAT_ENV_POSTGRES_URL`, the helper 
+will take that variable, convert it to the system property `postgres.url` and append it to the environment variable `JAVA_TOOL_OPTIONS`.
+
+Pseudocode:
+```
+	systemProperty = envVar.removePrefix('BPL_TOMCAT_ENV_').toLowerCase().replaceAll('_', '.')
+```
```

Comment:
This is the one I was thinking of: https://github.com/paketo-buildpacks/azure-application-insights/blob/main/helper/properties.go

The implementation here would have to be a bit different. Taking the key/values and appending them to `conf/catalina.properties`. The `$CATALINA_BASE` env should be set though, so you can read that to get the path to the config file, open and append to the end of it. I don't think it would be different for a value or for a file path. Unless you're trying to take the contents of the file and put that into catalina.properties, not just the path. That's feasible, just a little more effort.

One thing we'll need to do in the buildpack at build time is changing the permissions on the conf/catalina.properties file to be group writable. It is possible to run buildpack images as a different user, and if that's done the helper won't be able to write to the `/layers` or `/workspace` directory. Alternatively, we could put that file in `/tmp/` and symlink it to the layer but either way, you can write to the file at runtime, and IMHO it seems cleaner to just change the permissions.

Reply:
@dmikusa-pivotal, got a bit further on this - the file that exists in the buildpack is in `$CATALINA_HOME/conf/` and this doesn't seem to read by anything.  The logic to load those properties [looks in $CATALINA_BASE] (https://github.com/apache/tomcat/blob/8e2aa5e45ce13388da62386e3cb1dbfa3b242b4b/java/org/apache/catalina/startup/CatalinaProperties.java) but that file doesn't exist.  So rather than making the file group writable, we would just need to copy it into `$CATALINA_BASE/conf` and modify it there?

Reply:
That sounds like a good approach to me. Just to make sure I understand, what you're proposing is:

At build time:
1. Make a temp directory
2. Copy `$CATALINA_HOME/conf/catalina.properties` to the temp dir
3. Write `$CATALINA_BASE/conf/catalina.config` which points to the temp dir

At launch time:
1. Read `$CATALINA_BASE/conf/catalina.config` to get the temp dir
2. Read from the binding and append values to `tmp-dir/catalina.properties`.

I'm 👍 with this.

Reply:
Yeah, thats pretty much it, I'm not familiar with the `BPI_` env variables so hadn't considered those - but will look into it.  I'll get the RFC updated to reflect this.

Reply:
There's nothing special about `BPI_` just a convention we use on the Java buildpacks for something that's internal. It's typically when you need the build pack at build time to pass something to a helper/exec.d proc at launch time. The `BPI_` env's are not included in the docs or intended to be user-facing config.

Reply:
I was looking at the most recent changes and it looks like even using a temp directory, we cannot get around the need to make the file group writable. The buildpack is going to write that file to `/tmp/whatever/catalina.properties` and the owner is going to be set to `cnb` user, but at runtime if the user is different it won't be able to update the file :( At least not without making it group writable.

It makes me think we should just simplify then and do something like:

At build time:

1. Copy `$CATALINA_HOME/conf/catalina.properties` to `$CATALINA_BASE/conf/catalina.properties`
2. Make `$CATALINA_BASE/conf/catalina.properties` group writable.

At launch time:

1. Read `$CATALINA_BASE`
2. Read bindings & append values to `$CATALINA_BASE/conf/catalina.properties`

It's not a huge difference, but we can get rid of the `BPI_` env variable and the extra setting `-Dcatalina.config` and it should function the same and have no difference in security.

What do you think?

Reply:
That makes it much simpler, I've amended the proposal to include this